src/object/traverse.ts

```diff
@@ -0,0 +1,199 @@
+      : {
+          value: null,
+          key: null,
+          parent: null,
+          parents: [],
+          path: [],
+          skipped: new Set(),
+          skip(obj) {
+            context.skipped.add(obj ?? context.value)
+          },
+        }
+  ) as TraverseContext & {
+    value: unknown
+    key: keyof any
+    parent: any
+    parents: object[]
+    path: (keyof any)[]
+    skipped: Set<unknown>
+  }
+
+  let ok = true
+
+  // This is called for each value in an object or iterable.
+  const visit = (value: unknown, key: keyof any): boolean => {
+    // Map entries are destructured into value and key.
+    if (context.parent.constructor === Map) {
+      ;[key, value] = value as [keyof any, unknown]
+    }
+
+    context.path.push(key)
+    const result = visitor(
+      (context.value = value),
+      (context.key = key),
+      context.parent,
+      context,
+    )
+
+    if (result === false) {
+      return (ok = false)
+    }
+
+    // Traverse nested plain objects and arrays that haven't been
+    // skipped and aren't a circular reference.
+    if (
+      value !== null &&
+      typeof value === 'object' &&
+      (isArray(value) || isPlainObject(value)) &&
+      !context.skipped.has(value) &&
+      !context.parents.includes(value)
+    ) {
+      traverse(value)
+
+      if (isFunction(result)) {
+        result()
+      }
+    }
+
+    context.path.pop()
+    return ok
+  }
+
+  const traverse = (parent: object, isRoot?: boolean): boolean => {
+    context.parents.push(parent)
+    context.parent = parent
+
+    if (isArray(parent)) {
+      // Use Array.prototype.forEach for arrays so that sparse arrays
+      // are efficiently traversed. The array must be cloned so it can
+      // be cleared if the visitor returns false.
+      parent.slice().forEach((value, index, values) => {
+        if (visit(value, index) === false) {
+          values.length = 0 // Stop further traversal.
+        }
+      })
+    }
+    // Allow an iterable (e.g. a Map or a Set) to be passed in as the
+    // root object.
+    else if (isRoot && isIterable(parent)) {
+      let index = 0
+      for (const value of parent) {
+        if (visit(value, index) === false) {
+          return false
+        }
+        index++
+      }
+    }
+    // Traverse the object's properties.
+    else {
+      for (const key of ownKeys(parent)) {
+        if (visit(parent[key as keyof object], key) === false) {
+          return false
+        }
+      }
+    }
+
+    context.parents.pop()
+    context.parent = last(context.parents)
+
+    return ok
+  }
+
+  // If this is a recursive call, it's possible that the root object
+  // was skipped earlier. Check for that here so the caller doesn't
+  // have to check for it.
+  return outerContext?.skipped.has(root) || traverse(root, true)
+}
+
+/**
+ * The visitor callback for the `traverse` function.
+ */
+export type TraverseVisitor = (
+  value: unknown,
+  key: keyof any,
+  parent: object,
```

Comment:
We already have parent inside context, so we can skip that argument at all: 

```js
 _.traverse(obj, (value, key, {parent}) => {
})
```

For me, that API looks nicer. What do you think?

Reply:
This was intentional. The idea is to support curried functions that expect standard iterator functions. By matching the signature of `Array.prototype.forEach` callbacks for example, we leave open opportunities for code reuse.